psh/test-gibber.py

```diff
@@ -0,0 +1,80 @@
+# Phoenix-RTOS
+#
+# phoenix-rtos-tests
+#
+# psh gibbering input testing
+#
+# Copyright 2021, 2022 Phoenix Systems
+# Author: Damian Loewnau
+#
+# This file is part of Phoenix-RTOS.
+#
+# %LICENSE%
+#
+
+import string
+import psh.tools.psh as psh
+import trunner.config as config
+
+
+EXPECTED = 'Unknown command!'
+EOT = '\x04'
+
+
+def assert_letters(p):
+    msg = 'Wrong output for lowercase letters'
+    psh.assert_cmd(p, string.ascii_lowercase, EXPECTED, msg)
+
+    msg = 'Wrong output for uppercase letters'
+    psh.assert_cmd(p, string.ascii_uppercase, EXPECTED, msg)
+
+    msg = 'Wrong output for all ascii letters'
+    psh.assert_cmd(p, string.ascii_letters, EXPECTED, msg)
+
+
+def assert_digits(p):
+    msg = 'Wrong output for all digits'
+    psh.assert_cmd(p, string.digits, EXPECTED, msg)
+
+
+def assert_punctuation(p):
+    msg = 'Wrong output for punctuation characters'
+    psh.assert_cmd(p, string.punctuation, EXPECTED, msg)
+
+
+def assert_unprintable(p):
+    for ascii_code in (*range(32), 127):
+        # ^C is terminating character for qemu, so we skip it
+        if ascii_code == 3 and 'qemu' in config.CURRENT_TARGET:
+            continue
+        # ^D terminates psh process, so the another psh process is launched
+        if ascii_code == 4:
+            psh.assert_exec(p, 'psh')
+        # FS character skipped, to prevent EOF in reader buffer
+        if ascii_code == 28:
+            continue
+        cmd = chr(ascii_code)
+        msg = f'Wrong output for character with the following ascii code: {ascii_code}'
+        psh.assert_prompt_after_cmd(p, cmd, msg)
+
+        # LF or ^D, double prompt is expected
+        if ascii_code == 4 or ascii_code == 10:
+            psh.assert_prompt(p)
+
+
+def assert_not_ascii(p):
+    for code in range(128, 256):
+        cmd = chr(code)
+        msg = f'Wrong output for character with the following code: {code}'
+        psh.assert_prompt_after_cmd(p, cmd, msg)
```

Comment:
Waiting for prompt after every characters takes a lot of time.

You could test all nonprintable characters without side effects at once:
```python
s = ''.join(chr(x) for x in sorted({*range(0, 32),*range(127, 256)} - {3, 4, 10, 28, 127}))
output = psh.assert_prompt_after_cmd(p, characters, msg)
assert output == '', f"Unexpected output for non-printable characters: {output!r}"
```
^ Code above assumes assert_prompt_after_cmd returnes output.

Reply:
Yeah, this solution is much faster. I think it will be better.